### PR TITLE
[20227] Prevent index overflow and correctly assert the end iterator in DataSharing

### PR DESCRIPTION
**Pull request by Mario-DL**
## Description

When a DataSharing compatible publisher and subscriber were launched at high frequency (publication interval set to `0`), it happened two different issues:
* The current `notified_end` of the segment iterator was being updated before being actually checked, making an `assert()` compare two different versions of this iterator.
* On the other hand, unrelated with the previous observation, a corner case of advancing twice the `next_payload_` counter could occur.

Apart from the included test, in order to reproduce the first issue it is enough to launch a `BasicConfigurationExample` with a `bounded` type and no sleep between`writes()`. Launching, killing and relaunching the subscriber should trigger the `assertion`.
For reproducing the second issue, launch the subscriber first and then the publisher, a `SIGSEV` will be triggered in `20 seconds` or so.

This PR includes two fixes that mitigate the problem.
The idea of the test is that if in 20 seconds we have not crashed by some of the above issues, we give the test as passed.

**Important note** It was noticed that enabling all the verbosity messages in the test, makes the test not work properly under `colcon` since the `timeout` notification of the `SubscriberModule.cpp` condition variable is never fired. It seems that the test stresses the cpu and the timeout notification is lost. However, if the `python3` is directly invoked from a terminal (outside colcon), then there is no problem even if all the logs are enabled.

@Mergifyio backport 2.12.x 2.11.x 2.10.x 2.6.x

## Contributor Checklist
- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- [X] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- **N/A** Any new/modified methods have been properly documented using Doxygen. 
- [X ] Changes are ABI compatible. 
- [X] Changes are API compatible. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [X] Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by Mario-DL** [comment]
@richiprosima please test linux

**Comment by Mario-DL** [comment]
@richiprosima please test mac

**Comment by Mario-DL** [comment]
@richiprosima please test discovery_server

**Comment by Mario-DL** [comment]
@richiprosima please test mac

File: test/dds/communication/SubscriberMain.cpp [hunk 1]
```diff
@@ -41,9 +41,11 @@ int main(
     bool notexit = false;
     bool fixed_type = false;
     bool zero_copy = false;
+    bool succeed_on_timeout = false;
```
Comment (EduPonz) [inline test/dds/communication/SubscriberMain.cpp:44]: Add to the arguments comment above, same for timeout

**Review by EduPonz — CHANGES_REQUESTED** [review]
NIT not requiring CI

**Comment by Mario-DL** [comment]
@richiprosima please test this

**Review by EduPonz — APPROVED** [review]

**Comment by EduPonz** [comment]
@richiprosima please test mac

**Comment by Mario-DL** [comment]
@richiprosima please test mac

**Comment by Mario-DL** [comment]
@richiprosima please test mac

**Comment by Mario-DL** [comment]
@richiprosima please test mac

**Comment by Mario-DL** [comment]
@richiprosima please test mac

**Comment by Mario-DL** [comment]
@richiprosima please test mac

**Comment by Mario-DL** [comment]
@richiprosima please test mac

**Comment by JesusPoderoso** [comment]
CI issues unrelated to the PR. Ready to merge!

**Comment by MiguelCompany** [comment]
@Mergifyio backport 2.12.x 2.10.x 2.6.x
